compiler/qsc_partial_eval/src/lib.rs

```diff
@@ -0,0 +1,711 @@
+            let (_, _, callable_decl) = self.eval_callee_expr(*callee_expr_id);
+            callable_decl
+                .name
+                .name
+                .to_string()
+                .eq("__quantum__rt__qubit_release")
+        } else {
+            false
+        }
+    }
+
+    fn resolve_call_args(&mut self, args_expr_id: ExprId) -> Vec<rir::Value> {
+        let store_args_expr_id = StoreExprId::from((self.get_current_package_id(), args_expr_id));
+        let args_expr = self.package_store.get_expr(store_args_expr_id);
+        if let ExprKind::Tuple(exprs) = &args_expr.kind {
+            let mut values = Vec::<rir::Value>::new();
+            for expr_id in exprs {
+                self.eval_classical_expr(*expr_id);
+                let current_scope = self.eval_context.get_current_scope();
+                let expr_value = current_scope.get_expr_value(*expr_id);
+                let literal = map_eval_value_to_rir_literal(expr_value);
+                values.push(rir::Value::Literal(literal));
+            }
+            values
+        } else {
+            self.eval_classical_expr(args_expr_id);
+            let current_scope = self.eval_context.get_current_scope();
+            let args_expr_value = current_scope.get_expr_value(args_expr_id);
+            let literal = map_eval_value_to_rir_literal(args_expr_value);
+            vec![rir::Value::Literal(literal)]
+        }
+    }
+}
+
+impl<'a> Visitor<'a> for PartialEvaluator<'a> {
+    fn get_block(&self, id: BlockId) -> &'a Block {
+        let block_id = StoreBlockId::from((self.get_current_package_id(), id));
+        self.package_store.get_block(block_id)
+    }
+
+    fn get_expr(&self, id: ExprId) -> &'a Expr {
+        let expr_id = StoreExprId::from((self.get_current_package_id(), id));
+        self.package_store.get_expr(expr_id)
+    }
+
+    fn get_pat(&self, id: PatId) -> &'a Pat {
+        let pat_id = StorePatId::from((self.get_current_package_id(), id));
+        self.package_store.get_pat(pat_id)
+    }
+
+    fn get_stmt(&self, id: StmtId) -> &'a Stmt {
+        let stmt_id = StoreStmtId::from((self.get_current_package_id(), id));
+        self.package_store.get_stmt(stmt_id)
+    }
+
+    fn visit_expr(&mut self, expr_id: ExprId) {
+        if self.error.is_some() {
+            return;
+        }
+
+        // If the expression can be classically evaluated, do it. Otherwise, generate instructions for it.
+        let current_package_id = self.get_current_package_id();
+        let store_expr_id = StoreExprId::from((current_package_id, expr_id));
+        let expr_generator_set = self.compute_properties.get_expr(store_expr_id);
+        let callable_scope = self.eval_context.get_current_scope();
+        let compute_kind = expr_generator_set
+            .generate_application_compute_kind(&callable_scope.args_runtime_properties);
+        if matches!(compute_kind, ComputeKind::Classical) {
```

Comment:
```suggestion
        if compute_kind == ComputeKind::Classical {
```

Reply:
Cannot use `==` for type.

Reply:
Oh, I see. This type and few others would have to impl `PartialEq` and `Eq`, which is doable but not important right now.